[Backward Compatibility][Spot] Avoid cluster leakage by ray yaml overwritten and reduce spot controller cost on AWS

Michaelvll opened the pull request:
It resubmits #1221 , and add backward compatibility for it.

Update (Oct 13th):
* this fixes https://github.com/skypilot-org/skypilot/issues/640, ensuring restarting stopped clusters would not create new duplicate ones
* this reduces spot controller cost by 20%

Our current default instance created on the AWS has the volume with the maximum IOPs, i.e. 16000. However, by default, the instance created on the AWS console only has 3000 IOPs by default.
Our volume costs (16000-3000) * 0.005 / 30 / 24 = $0.09 / hour, which is 20% of the cost for our default CPU machine on AWS m6i.2xlarge.

Todo item: make this IOPs configurable across the clouds (probably simplify it by only having three tiers: low, medium, high).

Tested:
- [x] `sky cpunode -c test-iops`
- [x] `sky launch --disk-size 1024`
- [x] `sky launch --disk-size 50`
- [x] `sky gpunode -c test-t4 --gpus t4`
- [x] `sky launch -c bk-iops --num-nodes 2` (before this PR); `sky stop bk-iops`; `sky start bk-iops` it starts the same 2 instances.

concretevitamin commented on sky/backends/backend_utils.py:
```diff
@@ -655,6 +664,27 @@ def _remove_multinode_config(
 
+def _restore_original_block_for_yaml(new_yaml: str, old_yaml: str,
+                                     key_names: Set[str]) -> str:
+    """Restore the original block with key_name recursively."""
```
nit: Expand on 'recursively' a bit?  

```
Replaces 'new' with 'old' for all keys in key_names.

This is done recursively:  ..
```
actually no clear description comes to me.  This func is worth an example.

Michaelvll replied:
Good point! I added a comment:
```
Replaces 'new' with 'old' for all keys in key_names.

The replacement will be applied recursively and only for the blocks
with the key in key_names, and have the same ancestors in both 'new'
and 'old' YAML tree.
```

I found the example will be a bit difficult to cover all the cases. Do you think the description is clear enough?

concretevitamin replied:
Sgtm.

concretevitamin commented on sky/backends/backend_utils.py:
```diff
@@ -109,12 +110,22 @@
 # Remote dir that holds our runtime files.
 _REMOTE_RUNTIME_FILES_DIR = '~/.sky/.runtime_files'
 
+_RAY_YAML_KEYS_TO_RESTORE_FOR_BACK_COMPATIBILITY = {'Ebs'}
```
Needs a comment.

Michaelvll replied:
Added. Thanks!

concretevitamin reviewed (commented):
Nice @Michaelvll!  Questions:

- Does the mechanism here also fix  https://github.com/skypilot-org/skypilot/issues/640? E.g., put "everything below and including file_mounts can and should be overwritten" into the key set.
- Would https://github.com/skypilot-org/skypilot/blob/master/tests/backward_comaptibility_tests.sh catch the back compat issue?  Thinking whether next time we can just enforce running that test to catch such issues more easily.

Michaelvll commented:
> Does the mechanism here also fix https://github.com/skypilot-org/skypilot/issues/640? E.g., put "everything below and including file_mounts can and should be overwritten" into the key set.

This is a great point! I added the keys used for calculating the `launch_hash` in the key set. Supposedly, the issue will be fixed.

> Would https://github.com/skypilot-org/skypilot/blob/master/tests/backward_comaptibility_tests.sh catch the back compat issue? Thinking whether next time we can just enforce running that test to catch such issues more easily.

I just updated the script so that it can capture the problem with unexpectedly launching another instance. The script needs some refactorization. Probably, we can do it in another PR?

concretevitamin commented on sky/backends/backend_utils.py:
```diff
@@ -733,8 +750,11 @@ def write_cluster_config(to_provision: 'resources.Resources',
     if os.path.exists(yaml_path):
-        with open(yaml_path, 'r') as f:
-            old_yaml_content = f.read()
+        if force_overwrite:
```
When would be the case that the cluster didn't exist but this file exists?  If this is an exceptional case to guard against, rename it to `keep_launch_fields_in_existing_config: bool`?

Michaelvll replied:
The configure yaml files may not be correctly deleted if an error happens during `sky.down`. I did find multiple yaml files in the folder `~/.sky/generated` not belonging to any existing cluster. 
Good point! Let me rename the variable.

concretevitamin reviewed (commented):
Awesome! We can update the PR title/desc to reflect

- this fixes #640, ensuring restarting stopped clusters would not create new duplicate ones
- this reduces spot controller cost by 20%

Michaelvll commented on sky/backends/backend_utils.py:
```diff
@@ -110,7 +110,18 @@
+# - keeping the auth is not enough becuase the content of the key file will be used
+#   for calculating the hash.
+# TODO(zhwu): Keep in sync with the fields used in https://github.com/ray-project/ray/blob/e4ce38d001dbbe09cd21c497fedd03d692b2be3e/python/ray/autoscaler/_private/commands.py#L687-L701
+_RAY_YAML_KEYS_TO_RESTORE_FOR_BACK_COMPATIBILITY = {
```
I just added the `cluster_name` in the key set. For the `resources` and `head_node_type`, I think it would be better to add them when we actually want to modify those fields and find them affecting backward compatibility in the future.
As far as I understand, the `max_workers`, `upscaling_speed` and `idle_timeout_minutes` won't affect the launch hash.

Michaelvll commented on sky/backends/backend_utils.py:
```diff
@@ -666,7 +677,12 @@ def _remove_multinode_config(
 
 def _restore_original_block_for_yaml(new_yaml: str, old_yaml: str,
```
Done. Thanks!

concretevitamin reviewed (approved):
LGTM - this is awesome to have @Michaelvll.

Michaelvll commented:
Tested:
- [x] `./tests/backward_compatibility_tests.sh`
- [ ] `./tests/run_smoke_tests.sh`
